test/utils/integration/integration.go

```diff
@@ -0,0 +1,233 @@
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	"testing"
+	"time"
+)
+
+const (
+	useExistingCluster       = false
+	attachControlPlaneOutput = false
+	testEnvStartDelay        = time.Minute
+	testEnvStartAttempts     = 10
+	BigTimeOut               = 40 * time.Second
+	SmallPollingInterval     = 1 * time.Second
+)
+
+// TestEnvironment provides mocked resources for integration tests.
+type TestEnvironment struct {
+	Context          context.Context
+	EnvTestInstance  *envtest.Environment
+	k8sClient        client.Client
+	K8sDynamicClient *dynamic.DynamicClient
+	Reconciler       *eventing.Reconciler
+	Logger           *zap.SugaredLogger
+	Recorder         *record.EventRecorder
+	TestCancelFn     context.CancelFunc
+}
+
+//nolint:funlen // Used in testing
+func NewTestEnvironment(projectRootDir string, celValidationEnabled bool) (*TestEnvironment, error) {
+	var err error
+	// setup context
+	ctx := context.Background()
+
+	// setup logger
+	sugaredLogger, err := test.NewSugaredLogger()
+	if err != nil {
+		return nil, err
+	}
+
+	testEnv, envTestKubeCfg, err := StartEnvTest(projectRootDir, celValidationEnabled)
+	if err != nil {
+		return nil, err
+	}
+
+	// add to Scheme
+	err = eventingv1alpha1.AddToScheme(scheme.Scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	// +kubebuilder:scaffold:scheme
+
+	k8sClient, err := client.New(envTestKubeCfg, client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		return nil, err
+	}
+
+	dynamicClient, err := dynamic.NewForConfig(envTestKubeCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	// setup ctrl manager
+	metricsPort, err := test.GetFreePort()
+	if err != nil {
+		return nil, err
+	}
+
+	ctrlMgr, err := ctrl.NewManager(envTestKubeCfg, ctrl.Options{
+		Scheme:                 scheme.Scheme,
+		Port:                   metricsPort,
+		MetricsBindAddress:     "0", // disable
+		HealthProbeBindAddress: "0", // disable
+	})
+	if err != nil {
+		return nil, err
+	}
+	recorder := ctrlMgr.GetEventRecorderFor("eventing-manager")
+
+	// setup reconciler
+	eventingReconciler := eventingcontroller.NewReconciler(
+		ctrlMgr.GetClient(),
+		ctrlMgr.GetScheme(),
+		sugaredLogger,
+		recorder,
+	)
+	if err = (eventingReconciler).SetupWithManager(ctrlMgr); err != nil {
+		return nil, err
+	}
+
+	// start manager
+	var cancelCtx context.CancelFunc
+	go func() {
+		var mgrCtx context.Context
+		mgrCtx, cancelCtx = context.WithCancel(ctrl.SetupSignalHandler())
+		err = ctrlMgr.Start(mgrCtx)
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	return &TestEnvironment{
+		Context:          ctx,
+		k8sClient:        k8sClient,
+		K8sDynamicClient: dynamicClient,
+		Reconciler:       eventingReconciler,
+		Logger:           sugaredLogger,
+		Recorder:         &recorder,
+		EnvTestInstance:  testEnv,
+		TestCancelFn:     cancelCtx,
+	}, nil
+}
+
+func (env TestEnvironment) TearDown() error {
+	if env.TestCancelFn != nil {
+		env.TestCancelFn()
+	}
+
+	// retry to stop the api-server
+	sleepTime := 1 * time.Second
```

Comment:
(nit) Maybe 3-5 seconds is good, just to give it a bit time to stop before retrying.

Reply:
Done